tariff/ngeso.go

```diff
@@ -0,0 +1,128 @@
+package tariff
+
+import (
+	"errors"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/cenkalti/backoff/v4"
+	"github.com/evcc-io/evcc/api"
+	"github.com/evcc-io/evcc/tariff/ngeso"
+	"github.com/evcc-io/evcc/util"
+	"github.com/evcc-io/evcc/util/request"
+	"golang.org/x/exp/slices"
+)
+
+type Ngeso struct {
+	mux            sync.Mutex
+	log            *util.Logger
+	regionId       string
+	regionPostcode string
+	data           api.Rates
+	updated        time.Time
+}
+
+var _ api.Tariff = (*Ngeso)(nil)
+
+func init() {
+	registry.Add("ngeso", NewNgesoFromConfig)
+}
+
+func NewNgesoFromConfig(other map[string]interface{}) (api.Tariff, error) {
+	var cc struct {
+		Region   string
+		Postcode string
+	}
+
+	if err := util.DecodeOther(other, &cc); err != nil {
+		return nil, err
+	}
+
+	if cc.Region != "" && cc.Postcode != "" {
+		return nil, errors.New("cannot define region and postcode simultaneously")
+	}
+
+	t := &Ngeso{
+		log:            util.NewLogger("ngeso"),
+		regionId:       cc.Region,
+		regionPostcode: cc.Postcode,
+	}
+
+	done := make(chan error)
+	go t.run(done)
+	err := <-done
+
+	return t, err
+}
+
+func (t *Ngeso) run(done chan error) {
+	var once sync.Once
+	client := request.NewHelper(t.log)
+	bo := newBackoff()
+
+	// Use national results by default.
+	var tReq ngeso.CarbonForecastRequest
+	tReq = ngeso.ConstructNationalForecastRequest()
+
+	// If a region is available, use that.
+	// These should never be set simultaneously (see NewNgesoFromConfig), but in the rare case that they are,
+	// use the postcode as the preferred method.
+	if t.regionId != "" {
+		tReq = ngeso.ConstructRegionalForecastByIDRequest(t.regionId)
+	}
+	if t.regionPostcode != "" {
+		tReq = ngeso.ConstructRegionalForecastByPostcodeRequest(t.regionPostcode)
+	}
+
+	// Data updated by ESO every half hour, but we only need data every hour to stay current.
+	for ; true; <-time.Tick(time.Hour) {
+		var carbonResponse ngeso.CarbonForecastResponse
+		if err := backoff.Retry(func() error {
+			var err error
+			carbonResponse, err = tReq.DoRequest(client)
+
+			var statusError *request.StatusError
```

Comment:
Sorry, that was probably bad advise (or not the shortest option). This is even shorter and we know that there are not intermediate functions wrapping our error type:

			if se, ok := err.(request.StatusError); ok && se.HasStatus(http.StatusMethodNotAllowed) {
				// Catch cases where we're sending completely incorrect data (usually the result of a bad region).
				return backoff.Permanent(se)
			}
			return err

Reply:
Isn't `errors.As` the way to do it for nested error safety? https://go.dev/blog/go1.13-errors

Reply:
It is. But we know it can't be nested here and its one line less. Not important at all. Being a fan of consistency- that's how its currently implemented in other places in the codebase.

Reply:
Goland's linter is sulky at me for not wrapping, but committed your suggestion 👍